File: Back_SpringBoot/TappedOut/src/main/java/com/jgl/TappedOut/service/UserService.java
package com.jgl.TappedOut.service;

import java.util.List;

import com.jgl.TappedOut.dto.UserCreateDTO;
import com.jgl.TappedOut.dto.UserResponseDTO;
import com.jgl.TappedOut.dto.UserSecurityDTO;
import com.jgl.TappedOut.dto.UserUpdateDTO;

/**
 * Interface to declare methods needed at {@link SportLevelServiceImpl}
 * 
 * @author Jorge García López
 * @version 1.0
 * @since 2025
 */
public interface UserService {
    List<UserResponseDTO> getAllUsers();
    List<UserResponseDTO> getUsersByType(Long typeId);
    List<UserResponseDTO> getUsersByGender(Long genderId);
    List<UserResponseDTO> getUsersByLocation(String country, String city);
    List<UserResponseDTO> searchUsers(String query);
    UserResponseDTO getUserById(Long id);
    UserResponseDTO getUserByDni(String dni);
    UserResponseDTO getUserByEmail(String email);
    UserResponseDTO createUser(UserCreateDTO userCreateDTO);
    UserResponseDTO updateUser(Long id, UserUpdateDTO userUpdateDTO);
    UserResponseDTO updateUserSecurity(Long id, UserSecurityDTO userSecurityDTO);
    void deleteUser(Long id);
}

File: Front_React/tappedout-front/src/service/userService.js
/* eslint-disable no-useless-catch */
import apiClient from "../api/apiClient";
import { toast } from "react-toastify";

const userService = {
    getAllUsers: async () => {
        try {
            const response = await apiClient.get('/user');
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getUsersByType: async (userTypeId) => {
        try {
            const response = await apiClient.get(`/user/type/${userTypeId}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getUsersByGender: async (genderId) => {
        try {
            const response = await apiClient.get(`/user/gender/${genderId}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getUsersByLocation: async (country, city) => {
        try {
            const response = await apiClient.get('/user/location/', {
                params: {country, city}
            });
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    searchUsers: async (query) => {
        try {
            const response = await apiClient.get('/user/search', {
                params: {query}
            });
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getUserById: async (id) => {
        try {
            const response = await apiClient.get(`/user/${id}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getUserByDni: async (dni) => {
        try {
            const response = await apiClient.get(`/user/dni/${dni}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getUserByEmail: async (email) => {
        try {
            const response = await apiClient.get(`/user/email/${email}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    createUser: async (userData) => {
        try {
            const response = await apiClient.post('/user', userData);
            toast.success('User created successfully.');
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    updateUser: async (id, userData) => {
        try {
            const response = await apiClient.put(`/user/${id}`, userData);
            toast.success('User updated successfully.');
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    updateUserSecurity: async (id, userSecurityData) => {
        try {
            const response = await apiClient.patch(`/user/${id}/security`, userSecurityData);
            toast.success('User security information updated successfully.');
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    deleteUser: async (id) => {
        try {
            console.log('ID TO DELETE:', id);
            const response = await apiClient.delete(`/user/${id}`);
            toast.success('User deleted successfully.');
            return response.data;
        } catch (error) {
            throw error;
        }
    },
};

export default userService;